### PR TITLE
OSOE-743: Cancel ongoing runs of long-running workflows under a PR when new changes are pushed, vol.2

### LINKED ISSUE #288: Cancel ongoing runs of long-running workflows under a PR when new changes are pushed (OSOE-743)
This is particularly interesting for long-running workflows (project/solution builds) to reduce resource consumption, but small things add up for the shorter ones (e.g., spell-checking) as well. Example: When a commit in a PR triggers the solution build and another commit is pushed while the previous one's build is still running, we can optionally cancel the now outdated build using [concurrency groups](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#concurrency):

```yaml
concurrency:
  group: ${{ inputs.cancel-in-progress-for-this-pr == 'true' && github.event_name == 'pull_request' && format('{0}-{1}', github.workflow, github.ref) || github.run_id }}
  cancel-in-progress: ${{ inputs.cancel-in-progress-for-this-pr }}

on:
  workflow_call:
    inputs:
      cancel-in-progress-for-this-pr:
        type: string
        default: 'false'
```

[Jira issue](https://lombiq.atlassian.net/browse/OSOE-743)

### PR DESCRIPTION
**Pull request by BenedekFarkas**
Fixes #288

[OSOE-743](https://lombiq.atlassian.net/browse/OSOE-743)

[OSOE-743]: https://lombiq.atlassian.net/browse/OSOE-743?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .github/workflows/build-and-test-orchard-core.yml [hunk 1]
```diff
@@ -10,8 +10,22 @@ on:
           are used.
 
     inputs:
+      cancel-in-progress-for-this-pr:
+        description: >
+          When set to "true", it will cancel the already running workflow for this pull request. See the concurrency
+          settings of the job for more details.
+        type: string
+        default: 'true'
+      parent-job-name:
```
Comment (Piedone) [inline .github/workflows/build-and-test-orchard-core.yml:19]: Automatically including `build-directory˙` and `build-solution-path` would make further configuration for the OSOCE and Hastlayer SDK builds (and any workflow running these for multiple solutions in the same workflow) unnecessary. So, having that if it fits in `group` would be better.

Then, while we won't (at the moment) need any further key customization, we can have a general `concurrency-key-suffix` or something parameter here, that you can use to customize the key, with the job name (added as a tip to the description) or anything else we don't think about now.

Same for `build-and-test-dotnet`.

**Review by Piedone — CHANGES_REQUESTED** [review]

File: .github/workflows/build-and-test-orchard-core.yml (hunk 1 above)
Comment (BenedekFarkas) [reply .github/workflows/build-and-test-orchard-core.yml:19]: I tried that (adding `build-directory` and `build-solution-path`) already and it won't fit. It barely fits if we don't include `parent-job-name` and we "minify" the formatting to squeeze out all the spaces, that's a no-go. But if there was a new parameter in the strategy matrix (which has to be included in the concurrency key at the moment), then we're back to square 1.
The problematic limitation (= the 400 character limit applied to the concurrency key even it's an expression, not just a literal string) might be lifted in the future though, but we don't if/when. I've opened a [discussion](https://github.com/orgs/community/discussions/78279) about it.

BTW I just found [this post](https://github.com/orgs/community/discussions/63579#discussioncomment-6905355) about declaring the concurrency key for the caller job, which works correctly without having to worry about the strategy matrix, but then re-usability goes out the window.

Also, running the same workflow multiple times from the parent workflow is an edge case. We now have clear, easy-to-follow instructions for workflow authoring to support concurrency and handling this edge case from the caller workflow is just a simple parameter to be added.

Comment (Piedone) [reply .github/workflows/build-and-test-orchard-core.yml:19]: I see, then we can't have both a default that covers the multi-build workflow use case and a configurable key. However, why not add a general `concurrency-key-suffix` instead? (I.e. just change the name and description.) I'm sure we'll discover in the future that there are more edge cases that need a further customized key, so we can allow configuration for that already.

Comment (BenedekFarkas) [reply .github/workflows/build-and-test-orchard-core.yml:19]: The documentation/description already explains what it is and how to use it, but you can you can still plug any value into it if you want. The reason why I named it as such, because job names have to be unique (and should describe what the job is for), so when you see an evaluated key, you can infer what went into producing that key (BTW the key segments are added in order of specificity, so it goes kind of in the middle), in this case which job in the parent workflow called the reusable workflow.

Comment (BenedekFarkas) [reply .github/workflows/build-and-test-orchard-core.yml:19]: OK, so I slept on it. :D
I wanted to rename it to `concurrency-key-differentiator`, but that won't fit with the build-and-test-orchard-core workflow if we keep its name in the key as is (which I think is also very useful to keep like that for clarity). I'll just finish these without further changes, but hopefully they'll fix that issue about the 400 character limit.
